### ISSUE TITLE
add `--force` flag for `job deploy`

### ISSUE THREAD

**Issue by huanjani** [aws/copilot-cli#4188]
Related:
https://github.com/aws/copilot-cli/pull/2736
https://github.com/aws/copilot-cli/issues/2597
https://github.com/aws/copilot-cli/issues/2521
https://github.com/aws/copilot-cli/issues/1400
https://gitter.im/aws/copilot-cli?at=6373e658b1a759643672d7e2

**Comment by huanjani** [comment]
Note: `copilot deploy --force` should force job deployments, too.

**Comment by Lou1415926** [comment]
What is the use case of `job deploy --force` 🤔 ? Is this achieving the same thing as `job run`?

**Comment by anandsmalwade** [comment]
> What is the use case of `job deploy --force` 🤔 ? Is this achieving the same thing as `job run`?

No , "job run" will run the job , --force should be used to force deploy the job with the same version of docker image, we are able to do it for the service as of now.

so in short something like below 

copilot job deploy --env dev --name <name of the job> --force

**Comment by Lou1415926** [comment]
@anandsmalwade Thanks for adding the information! 

I wonder what are the behaviors you are seeing now? For example, suppose your job has been deployed using the manifest:
```yaml
image: 
   location: image:latest # `image:latest` is now on commit # 1.

schedule:
   on: 2m
```
Then if you:
1. Change the code and create commit # 2.
2. Push the docker image with `image:latest` tag - `image:latest` is now on commit # 2.
3. Wait for the next run of job to happen

Are the tasks that are kicked off *after* 2 still running commit # 1?

**Comment by anandsmalwade** [comment]
@Lou1415926  , I cannot push or deploy the docker image with commit #2 as the cloudformation stack will not detect any changes if I use the same docker image tag or version. so I will be stuck at step 2.

I see below message :

- No new infrastructure changes for stack <stackname>

![image](https://user-images.githubusercontent.com/17177050/205351566-b68caeb1-d51b-4d2b-b1fe-8f8f040d347a.png)

**Comment by Lou1415926** [comment]
@anandsmalwade I see. Your manifest looks like:
```yaml
image:
  build: path/to/Dockerfile
```
and you are running `copilot job deploy --env dev --name job --tag <imagetag>`, and `<imagetag> never changes - are these correct?

My thought is that, even though the CFN gives you error, Copilot should have successfully pushed the image to ECR with your tag. Since the tag is the same as the deployed template, the next execution of the job should pull image from <image:imagetag> which should point to your commit # 2. What you mind verifying this?

---

I have this thought because each execution of the job is under the hood, the statemachine kicking off a new ECS task. When ECS tries to spin up the task, it tries to pull the image from ECR. So the pulling happens every time a new task is started.

What happens with service is that - the tasks are usually long-running. Even if you push a new image using `docker push`, you still have to trigger the "kicking off of new tasks" in order for ECS to pull the new image. `svc deploy --force` is doing this - it is calling `UpdateService` API with `ForceNewDeployment` to let ECSs knows that I want to "kick off new tasks and replace the existing ones".

With job though - the state machine is naturally starting new task, so the trigger to pull image does not need to be pulled by `--force`.

In this case, I think the enhancement could probably become:
1. Run `job deploy --force`
2. Copilot push the artifacts, including the image as usual
3. Copilot deploys the CFN
4. If CFN complains about "no change", Copilot swallows that error instead of surfacing.

**Comment by anandsmalwade** [comment]
@Lou1415926  thanks for the detailed explanation ! 

my manifest looks like below 

```
image: 
   location: image:latest   # `image:latest` is now on commit # 1.

schedule:
   on: 2m
```

and yes I am deploying it with below 

copilot job deploy --env dev --name job --tag <imagetag>  and it ` never changes`


as per your explanation , I will verify from my end , if I see the expected changes in the  job even if I see the CFN error.

**Comment by Lou1415926** [comment]
@anandsmalwade Thanks for sharing the manifest!

Is that correct that, after you update your application code, you `docker push` the new image? 

If this is the case, you don't need to run `job deploy` every time after `docker push`. The next execution of the job should be able to pick up the new image that you pushed via `docker push`.

**Comment by anandsmalwade** [comment]
yes , after the code is committed to the repository , we build the docker image using Team City and then push the image to ECR in aws.

thanks a lot @Lou1415926 for all your help !
